src/Hazelcast.Net/FlakeId/AutoBatcher.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Hazelcast.Core;
+
+namespace Hazelcast.FlakeId
+{
+    internal class AutoBatcher
+    {
+        private readonly Func<Task<Batch>> _supplier;
+
+        private Lazy<Task<Batch>> _nextBatchLazyTask;
+
+        public AutoBatcher(Func<Task<Batch>> supplier)
+        {
+            _supplier = supplier;
+            _nextBatchLazyTask = NewBatchLazyTask();
+        }
+
+        public async Task<long> GetNextIdAsync(CancellationToken cancellationToken = default)
```

Comment:
I am wondering whether we could make this return `ValueTask<long>` and non-async for perfs purposes. Pseudo-code:
```
public ValueTask<long> GetNextIdAsync(...)
{
  if (we have a batch and a value) return new ValueTask(value); // this is fast
  return GetNextIdAsyncSlow(...);
}
private async ValueTask GetNextIdAsyncSlow(...)
{
  // go fetch a new batch etc
}
```

Reply:
I think you preferred `Task` for public API?

Reply:
When we are 100% certain that the operation will be async, yes a `Task` is preferred, but here we have to opportunity to optimize by avoiding the `async` state machine entirely. So I think a `ValueTask` is justified here.